crates/driver/src/simulator/mod.rs

```diff
@@ -0,0 +1,62 @@
+use {
+    crate::logic::{competition::solution, eth},
+    thiserror::Error,
+};
+
+mod node;
+mod tenderly;
+
+#[derive(Debug, Error)]
+#[error("simulation error")]
+pub struct Error;
+
+/// Ethereum transaction simulator.
+#[derive(Debug)]
+pub struct Simulator(Inner);
+
+impl Simulator {
+    /// Simulate an onchain settlement on the Ethereum network.
+    pub async fn simulate(
+        &self,
+        _settlement: &solution::Settlement,
+    ) -> Result<eth::Simulation, Error> {
+        // TODO This should do the two-step access list generation and gas estimation
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+enum Inner {
+    Tenderly {
+        tenderly: tenderly::Tenderly,
+        node: node::Node,
+    },
+    Node(node::Node),
+}
+
+impl Inner {
+    async fn access_list(&self, settlement: &solution::Settlement) -> eth::AccessList {
+        // The driver provides the users with an option of doing access list simulation
+        // on Tenderly because the eth_createAccessList endpoint seems to not be
+        // very widely supported yet. When that endpoint becomes more common, it
+        // might make sense to drop Tenderly entirely.
+        match self {
+            Self::Tenderly { tenderly, .. } => tenderly.access_list(settlement).await,
+            Self::Node(node) => node.access_list(settlement).await,
+        }
+    }
+
+    async fn gas(
+        &self,
+        settlement: &solution::Settlement,
+        access_list: &eth::AccessList,
+    ) -> eth::Gas {
+        // Gas estimation is always done using the node because it's faster than
+        // Tenderly and the driver has a connection to a node anyway.
```

Comment:
For the future, and not here - I would actually think that:
- Tenderly is just as fast, and maybe even faster with the `fast` flag
- Tenderly simulations allow for a "save_on_failure" flag to be set, so we automatically store failed simulations in the project's dashboard which would be **amazing** for debugging.

I think, for this PR it is fine to keep it like this, but as an additional "nice to have" feature, we should support simulations on Tenderly as well.

Reply:
Wow, didn't know that `tenderly` could be faster than using the node for simulation (seems like more overhead). But I guess it makes sense. They probably have their whole infrastructure optimized for running simulations.

Reply:
It isn't with the "full" flag - (since it does deeper tracing, and some source mapping), but the "fast" flag skips this and only does the simulation with the trace (which is a pretty small overhead in general - most nodes are IO bound on state lookup which Tenderly implements very well from what I understand - and anecdotally experienced).

Reply:
Either way, even it it is slower, it should be in the same order of magnitude, and since this simulation is done once I think the advantages (more convenient inspection in the Tenderly dashboard) completely outweigh the negative.